Ginkgo : Support K8s CI Coverage for Kafka GSG

manalibhutiyani opened the pull request:
These changes add support for testing Kafka GSG on a K8s environment in
Ginkgo. They follow the same steps as GSG and in addition a couple of other tests like policy trace,
to test validity of Kafka L7 policy enforcement on K8s environment.

Fixes Issue: #2602
Signed-Off-By: Manali Bhutiyani <manali@covalent.io>


```release-note
Ginkgo : Support K8s CI Coverage for Kafka GSG
```

Commit:
Ginkgo : Support multinode K8s CI Coverage for Kafka GSG
These changes add support for testing Kafka GSG on a multinode K8s environment in
Ginkgo. They follow the same steps as GSG and in addition a couple of other tests like policy trace,
to test validity of Kafka L7 policy enforcement on K8s multinode environment.

Fixes Issue: #2602
Signed-Off-By: Manali Bhutiyani <manali@covalent.io>

aanm reviewed (commented):
How are the tests multinode if the nodeSelector specified is the same one?

ianvernon commented on test/k8sT/KafkaPolicies.go:
```diff
@@ -0,0 +1,257 @@
+	"github.com/sirupsen/logrus"
+)
+
+var _ = Describe("K8sKafkaPolicyTest", func() {
```
If you want this to run as part of all PR runs, please change the name to `K8sValidatedKafkaPolicyTest`.

manalibhutiyani replied:
done

ianvernon commented on test/k8sT/KafkaPolicies.go:
```diff
@@ -0,0 +1,257 @@
+		kubectl = helpers.CreateKubectl(helpers.K8s1VMName(), logger)
+		podFilter = "k8s:zgroup=kafkaTestApp"
+
+		helpers.Sleep(5)
```
Leftover sleep? Please remove, as sleep calls cover-up timing problems with tests.

manalibhutiyani replied:
These were part of the many efforts to get the kube-dns part working.

manalibhutiyani replied:
done

ianvernon commented on test/k8sT/KafkaPolicies.go:
```diff
@@ -0,0 +1,257 @@
+				"cilium service list"})
+		}
+
+		kubectl.Delete(demoPath)
```
Deleting here isn't enough because calls to `kubectl delete` are asynchronous. Please add a call to `WaitCleanAllTerminatingPods` after this to wait for the result of `Delete` to succeed.

manalibhutiyani replied:
done

ianvernon commented on test/k8sT/KafkaPolicies.go:
```diff
@@ -0,0 +1,257 @@
+
+	})
+
+	waitUntilEndpointUpdates := func(pod string, eps map[string]int64, min int) error {
```
This function duplicates the one in `test/k8sT/Policies.go`. Please factor it out into a separate function used by both tests to avoid code duplication.

manalibhutiyani replied:
done

ianvernon commented on test/k8sT/KafkaPolicies.go:
```diff
@@ -0,0 +1,257 @@
+		return err
+	}
+
+	getAppPods := func() map[string]string {
```
While not exactly the same, this is quite similar to `test/k8s/Policies.go:getAppPods`. Factor this out into a function that both tests can use.

manalibhutiyani replied:
done

ianvernon commented on test/k8sT/KafkaPolicies.go:
```diff
@@ -0,0 +1,257 @@
+		return appPods
+	}
+
+	It("KafkaPolicies", func() {
```
Please rename this `It` to refer to the getting started guide.

ianvernon commented on test/k8sT/KafkaPolicies.go:
```diff
@@ -0,0 +1,257 @@
+	}
+
+	It("KafkaPolicies", func() {
+		clusterIP, err := kubectl.Get(helpers.DefaultNamespace, "svc").Filter(
```
Please utilize `By` throughout the test to describe the operations being performed.
For instane, on line 116, add `By("Getting Cilium Pods")`. This provides a useful narration to someone who doesn't know the ins-and-outs of the test if it fails.

eloycoto replied:
I would change this function to `kubectl.GetServiceHostPort`

ianvernon commented on test/k8sT/KafkaPolicies.go:
```diff
@@ -0,0 +1,257 @@
+		clusterIP, err := kubectl.Get(helpers.DefaultNamespace, "svc").Filter(
+			"{.items[?(@.metadata.name == \"kafka-service\")].spec.clusterIP}")
+		logger.Infof("PolicyRulesTest: cluster service ip '%s'", clusterIP)
+		Expect(err).Should(BeNil())
```
`Should` takes a second parameter which is a description string. Please add descriptions to each `Should` to provide context if the test fails.

manalibhutiyani replied:
This is similar to what is currently implemented for HTTP in policies.go. I will try adding more description.

ianvernon commented on test/k8sT/KafkaPolicies.go:
```diff
@@ -0,0 +1,257 @@
+
+		status := kubectl.CiliumExec(ciliumPod, fmt.Sprintf("cilium config %s=%s", helpers.PolicyEnforcement, helpers.PolicyEnforcementDefault))
+		status.ExpectSuccess()
+		helpers.Sleep(5)
```
Please remove calls to sleep; the following call to `CiliumEndpointWait` should take care of waiting for any necessary state changes in Cilium.

manalibhutiyani replied:
These were part of the many efforts to get the kube-dns part working.

tgraf replied:
Is it still needed?

manalibhutiyani replied:
no. not needed now. I am removing all the sleep calls

ianvernon commented on test/k8sT/KafkaPolicies.go:
```diff
@@ -0,0 +1,257 @@
+		Expect(endpoints.AreReady()).Should(BeTrue())
+
+		appPods := getAppPods()
+		By("Testing basic kafka setup")
```
Be more descriptive - what exactly constitutes a 'basic Kafka setup'? The GSG says 'Basic Kafka Produce and Consume' - would be helpful to correspond naming with the sections in the guide.

manalibhutiyani replied:
This is also meant to be Kafka K8s tests, not just GSG. Hence being generic in descriptions. I can change this, but the idea is that this test does more than  GSG

ianvernon commented on test/k8sT/KafkaPolicies.go:
```diff
@@ -0,0 +1,257 @@
+		appPods := getAppPods()
+		By("Testing basic kafka setup")
+
+		_, err = kubectl.ExecKafkaPodCmd(
```
If this first return value is always ignored, then remove it from `ExecKafkaPodCmd` and just return the error.

manalibhutiyani replied:
done

ianvernon commented on test/k8sT/KafkaPolicies.go:
```diff
@@ -0,0 +1,257 @@
+		appPods := getAppPods()
+		By("Testing basic kafka setup")
+
+		_, err = kubectl.ExecKafkaPodCmd(
```
Use `By` before these calls to `KafkaExecPodCmd` to narrate what is going on in the test.

ianvernon commented on test/k8sT/KafkaPolicies.go:
```diff
@@ -0,0 +1,257 @@
+		path = kubectl.ManifestGet("cilium_ds.yaml")
+		kubectl.Apply(path)
+		status, err := kubectl.WaitforPods(helpers.KubeSystemNamespace, "-l k8s-app=cilium", 300)
+		Expect(status).Should(BeTrue())
```
Add descriptions as strings to each call to `Should`, which will be displayed to users if the test fails.

manalibhutiyani replied:
This code is similar to whats done for Policy.go.

ianvernon commented on test/k8sT/KafkaPolicies.go:
```diff
@@ -0,0 +1,257 @@
+
+		kubectl.CiliumEndpointWait(ciliumPod)
+
+		epsStatus := helpers.WithTimeout(func() bool {
```
why not use `test/helpers/kubectl.go:CiliumEndpointWait` instead?

manalibhutiyani replied:
This code is similar to whats done for Policy.go.

ianvernon commented on test/k8sT/KafkaPolicies.go:
```diff
@@ -0,0 +1,257 @@
+
+		err = waitUntilEndpointUpdates(ciliumPod, eps, 6)
+		Expect(err).Should(BeNil())
+		epsStatus = helpers.WithTimeout(func() bool {
```
This duplicates the function above - factor it into something that is not duplicated, or alternatively, use `test/helpers/kubectl.go:CiliumEndpointWait`.

manalibhutiyani replied:
done

ianvernon commented on test/k8sT/KafkaPolicies.go:
```diff
@@ -0,0 +1,257 @@
+
+		_, err = kubectl.ExecKafkaPodCmd(
+			helpers.DefaultNamespace, appPods[helpers.EmpireHq], fmt.Sprintf("-c \"echo “Happy 40th Birthday to General Tagge” | ./kafka-produce.sh --topic empire-announce\""))
+		Expect(err).Should(BeNil())
```
How come the order of the produce / consume is different than in the GSG? In the GSG, the clients start to consume before the producer produces `empire-announce`.

manalibhutiyani replied:
We need to produce first, since consumer script *waits* for some messages to be already there by the producer. This test will hang if we consume first, since it will never return. This order is fine for this test

ianvernon replied:
OK, thanks; can you add a comment to the code indicating why the setup is performed in such a manner?

manalibhutiyani replied:
sure

manalibhutiyani replied:
done. added a comment to be more clear.

ianvernon commented on test/helpers/cons.go:
```diff
@@ -75,16 +75,20 @@ const (
+	Zookeeper = "zook"
+	Backup    = "empire-backup"
+	EmpireHq  = "empire-hq"
+	Outpost   = "empire-outpost"
```
why only outpost and not outpost-8888 and outpost-9999 as in the GSG?

eloycoto replied:
I would move this Backup, EmpireHq and Outpost variables to test. It's only used by the test and I don't think a test case that can be used too.

manalibhutiyani replied:
It doesn't make a difference because the policies applied to both outpost-8888 and outpost -9999 is same. They merely behave as two separate visual terminals for the GSG, but are exactly the same in behavior.

manalibhutiyani replied:
@eloycoto We have variables specific to HTTP tests defined here too. I can move it, but when we decide to add more Kafka tests, we might end up moving it again.

eloycoto replied:
If we add more kafka test it'll be added on `test/k8sT/KafkaPolicies.go` so no reason to have it on cons. 

The HTTP cons are used in `helpers/wrappers.go` and multiple runtime and k8st tests

manalibhutiyani replied:
Yes. My point is just like HTTP, Kafka needs to have more test coverage, and will have it eventually.  We will end up using these more as the Kafka tests are extended. Hence I think this is a good generic place for it.  Anyways, this is a trivial issue to debate over. I will change it.

ianvernon commented on test/k8sT/KafkaPolicies.go:
```diff
@@ -0,0 +1,257 @@
+		Expect(err).Should(BeNil())
+
+		_, err = kubectl.ExecKafkaPodCmd(
+			helpers.DefaultNamespace, appPods[helpers.Outpost], fmt.Sprintf("-c \"echo “Vader Booed at Empire Karaoke Party” | ./kafka-produce.sh --topic empire-announce\""))
```
Can you create wrappers around these `kafka-produce` and `kafka-consume` commands? There is lots of hardcoding of strings which can be factored into little helper functions.

eloycoto commented on test/helpers/kubectl.go:
```diff
@@ -71,6 +71,26 @@ func CreateKubectl(vmName string, log *logrus.Entry) *Kubectl {
+// The kafka producer and consumer scripts do not return error if command
+// leads to TopicAuthorizationException or any other error. Hence the
+// function needs to also take into account the stderr messages returned.
+func (kub *Kubectl) ExecKafkaPodCmd(namespace string, pod string, arg string) (string, error) {
```
I would move this function to the test itself. It's a specific function for one test case, and it doesn't make sense to me to have on the helpers. IMHO

eloycoto commented on test/k8sT/KafkaPolicies.go:
```diff
@@ -0,0 +1,257 @@
+			kubectl.CiliumReport(helpers.KubeSystemNamespace, ciliumPod, []string{
+				"cilium bpf tunnel list",
+				"cilium endpoint list",
+				"cilium service list"})
```
Here I would add `cilium policy get` too.

manalibhutiyani replied:
sure

manalibhutiyani replied:
done

eloycoto reviewed (changes requested):
The overall looks good, a few minor nitpicks plus the changes that Ian requested.

Thanks!

aanm commented on test/helpers/cons.go:
```diff
@@ -1,4 +1,4 @@
-// Copyright 2017 Authors of Cilium
+// Copyright 2018 Authors of Cilium
```
it should be `2017-2018`

manalibhutiyani replied:
done

Commit:
Ginkgo : Refactoring: K8s CI Coverage for Kafka GSG
   This change refactors all the common utility functions between HTTP tests in Policy.go, Nightly.go and KafkaPolicy.go
    Fixes Issue: #2602
    Signed-Off-By: Manali Bhutiyani <manali@covalent.io>

Commit:
Ginkgo : Move Kafka specific helpers in KafkaPolicies.go
 Fixes Issue: #2602
 Signed-Off-By: Manali Bhutiyani <manali@covalent.io>

Commit:
Add support for K1.6 ginkgo tests.
Fixes Issue: #2602
Signed-Off-By: Manali Bhutiyani <manali@covalent.io>

ianvernon reviewed (approved):
LGTM
